### PR TITLE
[FSDP] Add device in pin_memory argument

### PR DESCRIPTION
**Pull request by jeejakp12**
Add device to pin_memory argument to support other backends like HPU

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @rohan-varma

### CODE REVIEW

**Comment by colesbury** [comment]
@mrshenli would you please review this?

**Review by awgu — DISMISSED** [review]

File: torch/distributed/fsdp/_flat_param.py [hunk 1]
```diff
@@ -1184,12 +1184,12 @@ def init_flat_param_attributes(self) -> None:
         flat_param._local_shard = flat_param.data
         if self._offload_params:
             # Pin the memory for faster H2D transfer
-            flat_param._local_shard = flat_param._local_shard.pin_memory()
+            flat_param._local_shard = flat_param._local_shard.pin_memory(device=self.device)
```
[the remark this answers is not in the document]
Comment (jeejakp12) [reply torch/distributed/fsdp/_flat_param.py:1187]: device argument for pin_memory is to tell which device the host memory has to be pinned. the device here should be other than CPU. ex: cuda, hpu etc. 
For a CPU tensor, we want to pin the memory for a GPU, so self.device should be GPU.

[the remark this answers is not in the document]
Comment (awgu) [reply torch/distributed/fsdp/_flat_param.py:1187]: Got it! Thanks for the explanation.

Could you fix lint? Then, I think this PR is good to land.

[the remark this answers is not in the document]
Comment (jeejakp12) [reply torch/distributed/fsdp/_flat_param.py]: Done

**Comment by jeejakp12** [comment]
@awgu  i have fixed the lint error.

File: torch/distributed/fsdp/_flat_param.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeejakp12) [reply torch/distributed/fsdp/_flat_param.py]: @awgu  fixed the lint error

**Review by awgu — APPROVED** [review]

**Comment by bsochack** [comment]
@awgu should this change be rebased? and then can it be merged?

**Comment by awgu** [comment]
@bsochack sounds good!

**Comment by bsochack** [comment]
@awgu, do you think this is a real issue? I saw this test failing in the another PR with very similar timeout.

**Comment by awgu** [comment]
@bsochack I think it might have been a flaky issue. I re-ran the test, and it is fine now.

**Comment by bsochack** [comment]
@awgu thank you

**Pull request closed without merging**